File: JAVA/232 - Implement Queue using Stacks.java
/*
 232. Implement Queue using Stacks (E)

 Approach 1 : two stacks
 Use two stacks A & B, A as queue style stack, B as helper for reverse the A
 stack to make it "queue style".
 Time: push() - O(2n), pop() - O(1), empty() - O(1), peek() - O(1)
 Space: O(2n)

*/

/* Approach 1 */
class MyQueue {
    Stack<Integer> stack;
    Stack<Integer> helper;

    /** Initialize your data structure here. */
    public MyQueue() {
        stack = new Stack<Integer>();
        helper = new Stack<Integer>();
    }
    
    /** Push element x to the back of queue. */
    public void push(int x) {
        // Reverse the stack
        while (!stack.isEmpty()) helper.push(stack.pop());
        helper.push(x);
        while (!helper.isEmpty()) stack.push(helper.pop());
    }
    
    /** Removes the element from in front of queue and returns that element. */
    public int pop() {
        return stack.pop();
    }
    
    /** Get the front element. */
    public int peek() {
        return stack.peek();
    }
    
    /** Returns whether the queue is empty. */
    public boolean empty() {
        return stack.isEmpty();
    }
}

File: C++/232-Implement_Queue_using_Stacks.cpp
/*
 232. Implement Queue using Stacks (E)
 
 Notes:
    Store values "upside down" in stack to simulate the queue.
    Use two stacks, "q" and "helper", when user call the Push function, push & pop all elements in "q" stack
    to "helper" stack. After push the new value to "q" stack, take back all elements in "helper" stack. 
*/

class MyQueue {
private:
    stack<int> q, helper;
    
public:
    /** Initialize your data structure here. */
    MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        while (!q.empty())
        {
            helper.push(q.top());
            q.pop();
        }
        q.push(x);
        while (!helper.empty())
        {
            q.push(helper.top());
            helper.pop();
        }
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        int val = peek();
        q.pop();
        return val;
    }
    
    /** Get the front element. */
    int peek() {
        return q.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return q.empty();
    }
};
